src/Covenant/Internal/KindCheck.hs

```diff
@@ -72,12 +79,42 @@ checkKinds' = \case
     0 -> traverse_ checkKinds' nev
     _ -> throwError $ HigherOrderConstructorArg compT -- no higher order arguments to ctors (makes life much easier, not that useful)
   BuiltinFlat {} -> pure ()
-  Datatype tn args -> do
-    DataDeclaration _ numVars _ <- lookupDeclaration tn
-    let numArgsActual = V.length args
-        numArgsExpected = review intCount numVars
-    unless (numArgsActual == numArgsExpected) $ throwError (IncorrectNumArgs tn numVars args)
-    traverse_ checkKinds' args
+  Datatype tn args ->
+    lookupDeclaration tn >>= \case
+      OpaqueData {} -> pure ()
+      DataDeclaration _ numVars _ _ -> do
+        let numArgsActual = V.length args
+            numArgsExpected = review intCount numVars
+        unless (numArgsActual == numArgsExpected) $ throwError (IncorrectNumArgs tn numVars args)
+        traverse_ checkKinds' args
 
 checkKinds :: ValT AbstractTy -> Maybe KindCheckError
 checkKinds = either Just (const Nothing) . runKindCheckM . checkKinds'
+
+-- Ensures that we don't have any *mutually* recursive datatypes (which we don't want b/c they mess with data encoding strategies)
+cycleCheck :: forall (a :: Type). (Ord a) => Map TyName (DataDeclaration a) -> Maybe KindCheckError
+cycleCheck decls = either Just (const Nothing) $ runKindCheckM go
+  where
+    go =
+      local (\_ -> KindCheckContext decls) $
+        traverse_ (cycleCheck' Set.empty) =<< asks (view (#kindCheckContext % to M.elems))
+
+{- This is a bit odd b/c we don't want to fail for auto-recursive types, so we need to be careful
+   *not* to mark the current decl being examined as "visited" until we've "descended" into the dependencies
+   (I think?)
+-}
+cycleCheck' :: forall (a :: Type). (Ord a) => Set TyName -> DataDeclaration a -> KindCheckM a ()
+cycleCheck' _ OpaqueData {} = pure ()
+cycleCheck' visited (DataDeclaration tn _ ctors _) = traverse_ (checkCtor visited tn) ctors
+  where
+    checkCtor :: Set TyName -> TyName -> Constructor a -> KindCheckM a ()
+    checkCtor vs tn' (Constructor _ args) = do
+      let allComponents = Set.toList . Set.unions $ everythingOf <$> V.toList args
+          -- every type constructor in any part of a constructor arg, *except* the tycon of the decl
+          -- we're examining, since autorecursion is fine/necessary
+          allTyCons = Set.filter (/= tn') . Set.fromList . mapMaybe (fmap fst . preview datatype) $ allComponents
```

Comment:
This does a lot of list-munging, which isn't ideal. Is the fact that `Set`s lack `mapMaybe` the issue?

Reply:
Yes -_- I went looking for some alternative and the `containers` maintainers said they don't want to add a `Set.mapMaybe` & to just do something like this instead.